File: core/src/io/github/michelfaria/breadprototype/MyContactListener.java
package io.github.michelfaria.breadprototype;

import com.badlogic.gdx.physics.box2d.*;
import io.github.michelfaria.breadprototype.fud.PlayerFeetUD;
import io.github.michelfaria.breadprototype.fud.SolidUD;
import io.github.michelfaria.breadprototype.fud.WandProjectileUD;
import io.github.michelfaria.breadprototype.strategy.BlockSpawner;
import io.github.michelfaria.breadprototype.strategy.TodoListAppender;
import io.github.michelfaria.breadprototype.util.Pair;

import static io.github.michelfaria.breadprototype.util.FixtureUtil.*;

public class MyContactListener implements ContactListener {

    private final BlockSpawner blockSpawner;
    private final TodoListAppender todoListAppender;

    public MyContactListener(BlockSpawner blockSpawner, TodoListAppender todoListAppender) {
        this.blockSpawner = blockSpawner;
        this.todoListAppender = todoListAppender;
    }

    @Override
    public void beginContact(Contact contact) {
        final Pair<Fixture> fs = getFixturePair(contact);
        playerGrounded(fs, true);
        wandProjectileCollision(fs);
    }

    @Override
    public void endContact(Contact contact) {
        final Pair<Fixture> fs = getFixturePair(contact);
        playerGrounded(fs, false);
    }

    @Override
    public void preSolve(Contact contact, Manifold oldManifold) {

    }

    @Override
    public void postSolve(Contact contact, ContactImpulse impulse) {
    }

    private void playerGrounded(Pair<Fixture> fs, boolean grounded) {
        final Pair<Fixture> m = getFixturePairMatchByClass(fs, PlayerFeetUD.class, SolidUD.class);
        if (m != null) {
            final PlayerFeetUD fud = (PlayerFeetUD) getFUD(m.a);
            if (grounded) {
                fud.getPlayer().incrementGrounded();
            } else {
                fud.getPlayer().decrementGrounded();
            }
        }
    }

    private void wandProjectileCollision(Pair<Fixture> fs) {
        final Pair<Fixture> m = getFixturePairMatchByClass(fs, WandProjectileUD.class, Object.class);
        if (m != null) {
            final WandProjectileUD fud = (WandProjectileUD) getFUD(m.a);
            todoListAppender.addTask(() -> {
                if (!fud.isDisposed()) {
                    fud.touched(m.b);
                }
            });
        }
    }
}

File: core/src/io/github/michelfaria/breadprototype/MyContactListener.kt
package io.github.michelfaria.breadprototype

import com.badlogic.gdx.physics.box2d.*
import io.github.michelfaria.breadprototype.fud.PlayerFeetFUD
import io.github.michelfaria.breadprototype.fud.SolidFUD
import io.github.michelfaria.breadprototype.fud.WandProjectileFUD
import io.github.michelfaria.breadprototype.strategy.BlockSpawner
import io.github.michelfaria.breadprototype.strategy.TodoListAppender
import io.github.michelfaria.breadprototype.util.FixtureUtil.getFUD
import io.github.michelfaria.breadprototype.util.FixtureUtil.getFixturePair
import io.github.michelfaria.breadprototype.util.FixtureUtil.getFixturePairMatchByClass
import io.github.michelfaria.breadprototype.util.Pair

class MyContactListener(private val blockSpawner: BlockSpawner,
                        private val todoListAppender: TodoListAppender) : ContactListener {

    override fun beginContact(contact: Contact) {
        val fixtures = getFixturePair(contact)
        playerGrounded(fixtures, true)
        wandParticleCollision(fixtures)
    }

    override fun endContact(contact: Contact) {
        val fixtures = getFixturePair(contact)
        playerGrounded(fixtures, false)
    }

    private fun playerGrounded(fixtures: Pair<Fixture>, b: Boolean) {
        val match = getFixturePairMatchByClass(fixtures, PlayerFeetFUD::class.java, SolidFUD::class.java)
        if (match != null) {
            val fud = getFUD(match.a) as PlayerFeetFUD
            fud.player.isGrounded = b
        }
    }

    private fun wandParticleCollision(fixtures: Pair<Fixture>) {
        val match = getFixturePairMatchByClass(fixtures, WandProjectileFUD::class.java, SolidFUD::class.java)
        if (match != null) {
            val fud = getFUD(match.a) as WandProjectileFUD
            todoListAppender.addUpdate {
                if (!fud.isProjectileDead) {
                    fud.killProjectile()
                    blockSpawner.spawnBlock(match.a.body.position.x, match.a.body.position.y)
                }
            }
        }
    }

    override fun preSolve(contact: Contact, oldManifold: Manifold) {

    }

    override fun postSolve(contact: Contact, impulse: ContactImpulse) {

    }
}
